Adds face-api demos

joeyklee opened the pull request:
<!-- MANY MANY THANKS FOR YOUR CONTRIBUTION. ML5 ❤️S YOU! -->
## → Submit changes to the relevant branch 🌲

`development`


## → Description 📝

- submitting a new feature 🆕 

Adds [face-api.js](https://github.com/justadudewhohacks/face-api.js?files=1) to ml5.js

see PR to ml5-libary: https://github.com/ml5js/ml5-library/pull/482


## → Screenshots 🖼
(Any relevant screenshots, sketches, or helpful concept diagrams - these are helpful for our release notes and for getting people excited about your super cool feature and/or fix!)

##  FaceApi_Video_Expressions
![Screen Shot 2019-06-04 at 16 13 36](https://user-images.githubusercontent.com/3622055/58992391-b44a2100-87b8-11e9-92b6-a096072fc52c.png)

## FaceApi_Video_Landmarks
<img width="1212" alt="Screen Shot 2019-06-05 at 17 50 06" src="https://user-images.githubusercontent.com/3622055/58993262-04c27e00-87bb-11e9-906e-df892e1449f5.png">


## FaceApi_Video_Landmarks_LocalModels
<img width="1257" alt="Screen Shot 2019-06-05 at 17 34 12" src="https://user-images.githubusercontent.com/3622055/58992316-7cdb7480-87b8-11e9-8cf6-ab208409a332.png">

joeyklee committed:
added face-expressions demo

joeyklee committed:
added classifyExpressionsMultiple

joeyklee committed:
rm model files and model path requirement

joeyklee committed:
added face landmarks example

joeyklee committed:
clean up example

joeyklee committed:
rm options in example

joeyklee committed:
added local models example

joeyklee committed:
changed function names

joeyklee committed:
updated examples to use .detect()

joeyklee commented:
Eek! I need to fix the way the face features and bounding box are scaled and rendered 

![Screen Shot 2019-06-06 at 11 35 12](https://user-images.githubusercontent.com/3622055/59047648-43efde00-8852-11e9-9caa-d879c0b6d762.png)

joeyklee commented:
Ah cool! so we need to expose the `.resizeResults()` method 

![Screen Shot 2019-06-06 at 12 03 17](https://user-images.githubusercontent.com/3622055/59048093-2bcc8e80-8853-11e9-8cc9-c01c092cca15.png)

joeyklee committed:
rm translation

joeyklee committed:
added image example of Frida

joeyklee committed:
added detectSingle

joeyklee commented:
Adds also example using a single image

![Screen Shot 2019-06-06 at 13 30 05](https://user-images.githubusercontent.com/3622055/59053360-46a50000-885f-11e9-8c56-27f392b51856.png)

shiffman commented on p5js/FaceApi/FaceApi_Image_Landmarks/sketch.js:
```diff
@@ -0,0 +1,116 @@
+    createCanvas(200, 200);
+    img.resize(width, height);
+
+    faceapi = ml5.faceApi(detection_options, modelReady)
```
Are these optional and are there defaults? Or always required? I might suggest losing the `with` or `Face` if it's still clear enough? An old rule of thumb we always had back in the golden Processing days was two words only for methods and variables in the API 🤓 

This might be too bold though:

```javascript
const detection_options = {
    landmarks: true,
    expressions: false,
    descriptors: false,
}
```

joeyklee replied:
@shiffman - thanks for the comments!

Re licensing: 
+ the face-api.js is MIT licensed: https://github.com/justadudewhohacks/face-api.js/blob/master/LICENSE  so we should be ok!

Re options:
+ Yes the options are optional 😂 - by default we return all the data, but it runs a bit slower that way. So selectively choosing which data to return increases performance. So in this case people have to specify in the options which data to turn "off". 
+ I kept the **with** in `withLandmarks" etc since it refers to the way that the face-api.js refers to adding those data as part of the face detections. I think it is ok to also leave out the "with", but maybe it is helpful to make that conceptual linkage across the libraries? Also semantically to say "detections with landmarks"?

joeyklee replied:
changed options to →

```
{
    withLandmarks: true,
    withExpressions: true,
    withDescriptors: true,
}
```

shiffman commented on p5js/FaceApi/FaceApi_Image_Landmarks/sketch.js:
```diff
@@ -0,0 +1,116 @@
+    image(img, 0,0, width, height)
+    if (detections) {
+        // console.log(detections)
+        drawBox(detections)
```
This is great, but similarly to PoseNet we might also consider examples that show using smaller bits of the data?

joeyklee replied:
@shiffman - Oh I'm not sure I'm familiar. Do you mean that PoseNet is compressing the images?

shiffman replied:
Sorry, I was not clear at all! Let me try again! When we first introduced PoseNet the example that draws all of the keypoints and all of the skeleton examples overwhelmed users so it was nice to have one that, say, just shows you how to draw a circle on the nose or both eyes. Is there something similar we could do here? Like a simpler example that just draws the mouth? And maybe calculates the distance between top and bottom of mouth and maps that to a color or sound?

joeyklee replied:
@shiffman - ah yes, I think this is what we were mentioning in https://github.com/ml5js/ml5-examples/pull/161#discussion_r291630778

shiffman commented on p5js/FaceApi/FaceApi_Image_Landmarks/sketch.js:
```diff
@@ -0,0 +1,116 @@
+}
+
+function drawBox(detections){
+    const {alignedRect} = detections;
```
The destructuring syntax will probably be unfamiliar to people coming from intro p5.

joeyklee replied:
Right! I can change these to the variable and dot syntax

shiffman commented on p5js/FaceApi/FaceApi_Image_Landmarks/sketch.js:
```diff
@@ -0,0 +1,116 @@
+
+function drawBox(detections){
+    const {alignedRect} = detections;
+    const {_x, _y, _width, _height} = alignedRect._box;
```
This is a little silly, but I find that underscore variable naming can scare off beginners, such a small thing, but it makes the code appear to have a more obscure technical quality than it really has.

joeyklee replied:
You're right though! I'll err on the side of verbose variable names here.

shiffman commented on p5js/FaceApi/FaceApi_Image_Landmarks/sketch.js:
```diff
@@ -0,0 +1,116 @@
+    push()
+    // mouth
+    beginShape();
+    mouth.forEach(item => {
```
I tend to use `for...of` in my beginner p5 teaching, not sure what will ultimately be more consistent with p5 once more ES6 syntax gets adopted.

joeyklee replied:
Here I just used a regular for loop for now.

shiffman commented on p5js/FaceApi/FaceApi_Image_Landmarks/sketch.js:
```diff
@@ -0,0 +1,116 @@
+    // mouth
+    beginShape();
+    mouth.forEach(item => {
+        vertex(item._x, item._y)
```
Same comment re: underscore. If we are encouraging users of ml5 to access these properties I would lose the underscore probably?

joeyklee replied:
@shiffman - the underscore _x and _y are part of the way face-api.js returns the data. We could go through and update the "_x" and "_y" in the returns? I worry about adding additional computation as well as changing the data structure too much in case people are using face-api in other contexts. Should we keep the  "_x" and "_y" or rather change it?

shiffman replied:
Ah, I see, I didn't realize this was the face API directly! I think we can keep it as is and maybe file this as something to think about later if it's causing confusion or we end up wrapping elements of the API for other reasons?

joeyklee replied:
@shiffman - I wonder now if maybe it just makes sense to include in the return:

```
landmarkFeatures:
{
mouth: [array of {_x, _y}]
nose:[array of {_x, _y}]
leftEye:[array of {_x, _y}]
rightEye: [array of {_x, _y}]
...etc
}
```

shiffman replied:
Oh, I like this idea! Very similar to PoseNet actually!

joeyklee replied:
Cool! I'll add this in now :)

shiffman replied:
What if we just call it `landmarks`? or `features`?

joeyklee replied:
Since `landmarks` is the name of the returned results of the landmarks, I'd go for maybe just "parts" ?

shiffman commented on p5js/FaceApi/FaceApi_Video_Expressions/sketch.js:
```diff
@@ -0,0 +1,56 @@
+
+    background(255);
+    if (detections) {
+        if (detections.length > 0) {
```
Is there a way we can simplify this for the user (this actually comes up with a lot of other ml5 functions, e.g. image classification where I am often writing `if (results && results.length > 0)`. Maybe we don't even trigger the callback if there are no results? Or we have some sort of property like `results.success`? This probably belongs in its own issue / discussion.

joeyklee replied:
@shiffman - oh this is definitely something for a larger discussion. Let's put a pin in this for now 📍

shiffman commented on p5js/FaceApi/FaceApi_Video_Expressions/sketch.js:
```diff
@@ -0,0 +1,56 @@
+    if (detections) {
+        if (detections.length > 0) {
+            const {expressions} = detections[0]
+            let keys = Object.keys(expressions);
```
Maybe we provide a way of doing this iteration internally? Could detections just have a sorted array, kind of like what we do with image classification results?

joeyklee replied:
@shiffman  - I think I would stay with the {object} here. If we return an array of objects with something like {label, confidence}, the issue is that we will always have to loop through and array to search for the expression of interest. Since we have a fixed number and the uses might be specific enough where people are only interested in "happy" or "surprised", do you think it would be best to keep it as an object rather than an array? 

```
{
happy: 0.9,
sad: 0.05,
surprised: 0.05,
etc
}
```

shiffman commented on p5js/FaceApi/FaceApi_Video_Expressions/sketch.js:
```diff
@@ -0,0 +1,56 @@
+        if (detections.length > 0) {
+            const {expressions} = detections[0]
+            let keys = Object.keys(expressions);
+            keys.forEach( (item, idx) => {
```
Regular for loop here with `i`?

joeyklee replied:
@shiffman - yes, we can switch to a regular for-loop.

shiffman commented on p5js/FaceApi/FaceApi_Video_Landmarks_LocalModels/sketch.js:
```diff
@@ -0,0 +1,124 @@
+    withFaceLandmarks: true,
+    withFaceExpressions: false,
+    withFaceDescriptors: false,
+    Mobilenetv1Model: 'models',
```
Are these almost always going to be the same? Maybe something like `modelPath: path` would be simpler?

joeyklee replied:
@shiffman - actually, in this case, by default we read in the data via URL. If people will be specifying a directory path to the varying models, it could be that people have different directories for different models. If we only have "modelPath", then we force people to always use 1 directory. I'd keep it a bit more open ended since almost 99% of the time people will use the pretrained models that are being read in via URL?

shiffman replied:
yes, that makes sense!!

shiffman reviewed (commented):
This is super exciting, thank you for this massive effort! 🤩

@joeyklee, do you think we should reach out to creator of face-api for thoughts / feedback? I assume the license is such that incorporating it as part of ml5 presents no issues? 

I am particularly excited by the idea of tying this to KNNClassification and a new generic Neural Network API! (https://github.com/ml5js/ml5-library/issues/483)

joeyklee committed:
updated image landmarks with landmarks.parts

joeyklee committed:
refactored code friendly and changed colors

joeyklee commented:
Added {result.parts} object with mouth, nose, etc: https://github.com/ml5js/ml5-library/pull/482/commits/7136aa05abda4dd70e720b247b5c07fe1311423d 

added the ml5 color purple for the stroke. 

![Screen Shot 2019-06-07 at 11 44 52](https://user-images.githubusercontent.com/3622055/59116654-0ce4ff80-891a-11e9-9d99-035ca2214b55.png)

joeyklee committed:
made ex friendly

joeyklee committed:
changed options from withFaceX to withX

joeyklee commented:
Merging now! Thanks for all the feedback @shiffman ! 🙏

thirukumars commented:
how to use the face-api with ml5js
